### ISSUE TITLE
Supertux crashes after resetting level right after throwing an ice cube

### ISSUE THREAD

**Issue by BlendyDev** [SuperTux/supertux#2362]
**Supertux version**: v0.6.3
**OS/Distro**: Arch Linux
**Kernel**: 6.0.12-arch1-1 (64-bit)
#### Expected behavior
Game doesn't crash
#### Actual behavior
Game crashes (segfault)
#### Steps to reproduce actual behavior
Play level two (The Journey Begins), after throwing an ice cube press esc and restart the level (don't know the exact cause, but this happened three times while speedrunning, and they were all on this situation)
#### Additional debugging information
```
Error: signal 11:
supertux2(_ZN12ErrorHandler17print_stack_traceEv+0x30)[0x5607c50ec490]
supertux2(_ZN12ErrorHandler12handle_errorEi+0x3c)[0x5607c50ec52c]
/usr/lib/libc.so.6(+0x38a00)[0x7f983f569a00]
supertux2(_ZN10GameObjectD2Ev+0x3d)[0x5607c50ed76d]
supertux2(_ZN10MrIceBlockD0Ev+0x178)[0x5607c5193c28]
supertux2(_ZN17GameObjectManager13clear_objectsEv+0x92)[0x5607c50f8d52]
supertux2(_ZN6SectorD2Ev+0x27)[0x5607c4fc3407]
supertux2(_ZN6SectorD0Ev+0xd)[0x5607c4fc358d]
supertux2(_ZN5LevelD2Ev+0x43)[0x5607c5103c73]
supertux2(_ZN11GameSession13restart_levelEb+0xea)[0x5607c50fbfca]
supertux2(_ZN11GameSession6updateEfRK10Controller+0x34c)[0x5607c50fd2ec]
supertux2(_ZN13ScreenManager16update_gamelogicEf+0x4e)[0x5607c4fba93e]
supertux2(_ZN13ScreenManager9loop_iterEv+0x18a)[0x5607c4fbc1ea]
supertux2(_ZN13ScreenManager3runEv+0x28)[0x5607c4fbc4b8]
supertux2(_ZN4Main3runEiPPc+0x4b3)[0x5607c4f80a73]
supertux2(main+0x68)[0x5607c4f507e8]
/usr/lib/libc.so.6(+0x23290)[0x7f983f554290]
/usr/lib/libc.so.6(__libc_start_main+0x8a)[0x7f983f55434a]
supertux2(_start+0x25)[0x5607c4f57cd5]
^[[C[1]    5823 segmentation fault (core dumped)  supertux2
```
```

           PID: 5823 (supertux2)
           UID: 1000 (blendy)
           GID: 1000 (blendy)
        Signal: 11 (SEGV)
     Timestamp: Sat 2022-12-17 12:30:31 CET (13min ago)
  Command Line: supertux2
    Executable: /usr/bin/supertux2
 Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/app-org.kde.konsole-e554f1c9e6cd454682730216489d6342.scope
          Unit: user@1000.service
     User Unit: app-org.kde.konsole-e554f1c9e6cd454682730216489d6342.scope
         Slice: user-1000.slice
     Owner UID: 1000 (blendy)
       Boot ID: ba2e1ffb6341445c85234d1ae73a664e
    Machine ID: 64c709bc748d49b2a97cda6779e4cc50
      Hostname: arch
       Storage: /var/lib/systemd/coredump/core.supertux2.1000.ba2e1ffb6341445c85234d1ae73a664e.5823.1671276631000000.zst (present)
  Size on Disk: 17.7M
       Message: Process 5823 (supertux2) of user 1000 dumped core.
                
                Stack trace of thread 5823:
                #0  0x00007f983fed7380 FT_Done_Face (libfreetype.so.6 + 0x18380)
                #1  0x00005607c52a44b2 n/a (supertux2 + 0x51a4b2)
                #2  0x00005607c4feb53c _ZN7TTFFontD0Ev (supertux2 + 0x26153c)
                #3  0x00005607c4fb5d21 _ZNSt10shared_ptrI4FontED2Ev (supertux2 + 0x22bd21)
                #4  0x00007f983f56bfa5 n/a (libc.so.6 + 0x3afa5)
                #5  0x00007f983f56c120 exit (libc.so.6 + 0x3b120)
                #6  0x00005607c50ec4e5 _ZN12ErrorHandler13close_programEv (supertux2 + 0x3624e5)
                #7  0x00005607c50ec531 _ZN12ErrorHandler12handle_errorEi (supertux2 + 0x362531)
                #8  0x00007f983f569a00 n/a (libc.so.6 + 0x38a00)
                #9  0x00005607c50ed76d _ZN10GameObjectD2Ev (supertux2 + 0x36376d)
                #10 0x00005607c5193c28 _ZN10MrIceBlockD0Ev (supertux2 + 0x409c28)
                #11 0x00005607c50f8d52 _ZN17GameObjectManager13clear_objectsEv (supertux2 + 0x36ed52)
                #12 0x00005607c4fc3407 _ZN6SectorD2Ev (supertux2 + 0x239407)
                #13 0x00005607c4fc358d _ZN6SectorD0Ev (supertux2 + 0x23958d)
                #14 0x00005607c5103c73 _ZN5LevelD2Ev (supertux2 + 0x379c73)
                #15 0x00005607c50fbfca _ZN11GameSession13restart_levelEb (supertux2 + 0x371fca)
                #16 0x00005607c50fd2ec _ZN11GameSession6updateEfRK10Controller (supertux2 + 0x3732ec)
                #17 0x00005607c4fba93e _ZN13ScreenManager16update_gamelogicEf (supertux2 + 0x23093e)
                #18 0x00005607c4fbc1ea _ZN13ScreenManager9loop_iterEv (supertux2 + 0x2321ea)
                #19 0x00005607c4fbc4b8 _ZN13ScreenManager3runEv (supertux2 + 0x2324b8)
                #20 0x00005607c4f80a73 _ZN4Main3runEiPPc (supertux2 + 0x1f6a73)
                #21 0x00005607c4f507e8 main (supertux2 + 0x1c67e8)
                #22 0x00007f983f554290 n/a (libc.so.6 + 0x23290)
                #23 0x00007f983f55434a __libc_start_main (libc.so.6 + 0x2334a)
                #24 0x00005607c4f57cd5 _start (supertux2 + 0x1cdcd5)
                
                Stack trace of thread 5830:
                #0  0x00007f983f639096 epoll_wait (libc.so.6 + 0x108096)
                #1  0x00007f983c0b5789 n/a (libspa-support.so + 0x14789)
                #2  0x00007f983c0a7b9b n/a (libspa-support.so + 0x6b9b)
                #3  0x00007f9838eb5177 n/a (libpipewire-0.3.so.0 + 0x42177)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                
                Stack trace of thread 5832:
                #0  0x00007f983f639096 epoll_wait (libc.so.6 + 0x108096)
                #1  0x00007f983c0b5789 n/a (libspa-support.so + 0x14789)
                #2  0x00007f983c0a7b9b n/a (libspa-support.so + 0x6b9b)
                #3  0x00007f9838ef4a4a n/a (libpipewire-0.3.so.0 + 0x81a4a)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                
                Stack trace of thread 5833:
                #0  0x00007f983f639096 epoll_wait (libc.so.6 + 0x108096)
                #1  0x00007f983c0b5789 n/a (libspa-support.so + 0x14789)
                #2  0x00007f983c0a7b9b n/a (libspa-support.so + 0x6b9b)
                #3  0x00007f9838ef4a4a n/a (libpipewire-0.3.so.0 + 0x81a4a)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                
                Stack trace of thread 5834:
                #0  0x00007f983f639096 epoll_wait (libc.so.6 + 0x108096)
                #1  0x00007f983c0b5789 n/a (libspa-support.so + 0x14789)
                #2  0x00007f983c0a7b9b n/a (libspa-support.so + 0x6b9b)
                #3  0x00007f9838eb5177 n/a (libpipewire-0.3.so.0 + 0x42177)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                
                Stack trace of thread 5836:
                #0  0x00007f983f5b44b6 n/a (libc.so.6 + 0x834b6)
                #1  0x00007f983f5bf6a0 n/a (libc.so.6 + 0x8e6a0)
                #2  0x00007f983fd1a85f n/a (libopenal.so.1 + 0x2b85f)
                #3  0x00007f983f8d62f3 execute_native_thread_routine (libstdc++.so.6 + 0xd62f3)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                
                Stack trace of thread 5846:
                #0  0x00007f983f5b44b6 n/a (libc.so.6 + 0x834b6)
                #1  0x00007f983f5b72d7 pthread_cond_clockwait (libc.so.6 + 0x862d7)
                #2  0x00005607c5311726 _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN14IoThreadHolder5StartEvEUlvE_EEEEE6_M_runEv (supertux2 + 0x587726)
                #3  0x00007f983f8d62f3 execute_native_thread_routine (libstdc++.so.6 + 0xd62f3)
                #4  0x00007f983f5b78fd n/a (libc.so.6 + 0x868fd)
                #5  0x00007f983f639a60 n/a (libc.so.6 + 0x108a60)
                ELF object binary architecture: AMD x86-64
```

**Comment by BlendyDev** [comment]
[Recorded crash](https://www.youtube.com/watch?v=I40HvLbYYRQ)

**Comment by Semphriss** [comment]
For any developer involved with this:

The exact reproduction steps are: Grab a cube, walk through any badguy so that both that badguy and the cube you carry die (don't ever throw the cube, let it touch the other badguy while still carrying it), before the cube falls of the screen press escape and restart the level, then restart the level a second time. The crash should happen on the second reset.

Where I am so far: The reason of the crash is because the ObjectRemoveListeners are not properly handled between the player and the ice cube upon destruction. I tried investigating why the code doesn't work, but when I tried tracing the code, the result was so messy that it probably uncovered a dozen different bugs. Will keep investigating later.
